File: Week2/c++/Deck.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <random>
#include <chrono>
#include <array>

using namespace std;

//Class representing a card
class Card{
	private:
		string suit;
		string rank;

	public:
		Card() : suit(""), rank(""){}
		Card(string s, string r) : suit(s), rank(r){}

		void set_suit(string s){suit = s;}
		void set_rank(string r){rank = r;}

		string get_suit(){return suit;}
		string get_rank(){return rank;}
};

//Class representing a deck
class Deck{
	private:
		array<Card,52> deck;
		int draw_amount = 0;
		string suits[4] = {"hearts","diamonds","spades","clubs"};
		string ranks[13] = {"2", "3", "4", "5", "6", "7", "8", "9", "10", "Jack", "Queen", "King", "Ace"};
	public:
		//When a deck is made, it is filled with 52 cards and shuffled
		Deck(){
			for(int i=0; i<4; i++){
				for(int j=0;j<13;j++){
					deck[i * 13 + j].set_suit(suits[i]);
					deck[i * 13 + j].set_rank(ranks[j]);
				}
			}
			unsigned seed = chrono::system_clock::now().time_since_epoch().count();
			shuffle(deck.begin(), deck.end(), default_random_engine(seed));
		}

		//Returns the current card from the stack and incriments the pointer by one
		Card draw_card(){
			Card card = deck[draw_amount];
			draw_amount++;
			return card;
		}
};

File: Week2/Python/Deck.py
import random

#Creates a new deck and shuffles it
class Deck:
    def __init__(self):
        self.suits = ["spades","clubs","diamonds","hearts"]
        self.ranks = [2,3,4,5,6,7,8,9,10,"Jack","Queen","King","Ace"]
        self.generate_deck()


    def shuffle(self):
        random.shuffle(self.deck)
    
    #Generates a new deck of 52 playing cars
    def generate_deck(self):
        self.deck = []
        for suit in self.suits:
           for rank in self.ranks:
                self.deck.append((rank,suit))
        self.shuffle()

    #Draws a card which is removed from the deck
    def draw_card(self):
        card = self.deck.pop()
        return card

